shortfin/python/shortfin_apps/llm/components/invocation.py

```diff
@@ -1,29 +1,78 @@
 import logging
 import math
+import traceback
 
 import shortfin as sf
 import shortfin.array as sfnp
 
 from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from itertools import chain
 from typing import List, Optional, Tuple, Union
 
 from .buffers import copy_buffers_to_host, create_argument_buffers
+from .kvcache.base_attention_cache import BasePagedAttentionCache
 from .device_array_cache import Allocation, DeviceArrayCache, WrappedAllocation
-from .messages import LlmInferenceExecRequest
+from .messages import LlmInferenceExecRequest, InferencePhase
 
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class LlmTaskInput:
+    block_count: int
+    seq_stride: int
+    input_tokens: List[List[int]]
+    page_ids: List[List[int]]
+
+    start_positions: Optional[List[int]] = None
+
+    @property
+    def batch_seq_len(self):
+        seq_stride = self.seq_stride
+        bsl = max(len(tokens) for tokens in self.input_tokens)
+        return int(math.ceil(bsl / seq_stride) * seq_stride)
+
+    @staticmethod
+    def from_exec_requests(
```

Comment:
This construction should be in `batcher` not `invocation`. If you are breaking the `exec_req` requirement we don't want to continue its use in `invocation.py`.